### PR TITLE
BUG: resample fixes

### LINKED ISSUE #12448: Breaking examples due to resample refactor
While using master a bit, I discovered some more cases where the new resample API breaks things:

- Plotting. `.plot` is a dedicated groupby/resample method (which adds each group individually to the plot), while I think it is a very common idiom to quickly resample your timeseries and plot it with (old API) eg `s.resample('D').plot()`. 
  Example with master:

  ```
In [1]: s = pd.Series(np.random.randn(60), index=date_range('2016-01-01', periods=60, freq='1min'))

In [3]: s.resample('15min').plot()
Out[3]:
2016-01-01 00:00:00    Axes(0.125,0.1;0.775x0.8)
2016-01-01 00:15:00    Axes(0.125,0.1;0.775x0.8)
2016-01-01 00:30:00    Axes(0.125,0.1;0.775x0.8)
2016-01-01 00:45:00    Axes(0.125,0.1;0.775x0.8)
Freq: 15T, dtype: object
  ```

  figure_1

  while previously it would just have given you one continuous line. 
  This one can be solved I think by special casing `plot` for `resample` (not have it a special groupby-like method, but let it warn and pass the the `resample().mean()` result to `Series.plot()` like the 'deprecated_valids')

- When you previously called a method on the `resample` result that is also a valid Resampler method now. Eg `s.resample(freq).min()` would previously have given you the "minimum daily average" while now it will give you the "minimum per day". 
  This one is more difficult/impossible to solve I think? As you could detect that case if you know it is old code, but cannot distinguish it from perfectly valid code with the new API. If we can't solve it, I think it deserves some mention in the whatsnew explanation.

- Using `resample` on a `groupby` object (xref #12202). Using the example of that issue, with 0.17.1 you get:

  ```
In [1]: df = pd.DataFrame({'date': pd.date_range(start='2016-01-01', periods=4,
freq='W'),
   ...:                'group': [1, 1, 2, 2],
   ...:                'val': [5, 6, 7, 8]})

In [2]: df.set_index('date', inplace=True)

In [3]: df
Out[3]:
          group  val
date
2016-01-03      1    5
2016-01-10      1    6
2016-01-17      2    7
2016-01-24      2    8

In [4]: df.groupby('group').resample('1D', fill_method='ffill')
Out[4]:
                    val
group date
1     2016-01-03    5
      2016-01-04    5
      2016-01-05    5
      2016-01-06    5
      2016-01-07    5
      2016-01-08    5
      2016-01-09    5
      2016-01-10    6
2     2016-01-17    7
      2016-01-18    7
      2016-01-19    7
      2016-01-20    7
      2016-01-21    7
      2016-01-22    7
      2016-01-23    7
      2016-01-24    8

In [5]: pd.__version__
Out[5]: u'0.17.1'
  ```

  while with master you get:

  ```
In [29]: df.groupby('group').resample('1D', fill_method='ffill')
Out[29]: <pandas.core.groupby.DataFrameGroupBy object at 0x0000000009BA73C8>
  ```

  which will give you different results/error with further operations on that. Also, this case does not raise any FutureWarning (which should, as the user should adapt the code to `groupby().resample('D').ffill()`)

### PR DESCRIPTION
**Pull request by jreback**
make sure .resample(...).plot() warns and returns a correct plotting object
make sure that .groupby(...).resample(....) is hitting warnings when appropriate

closes #12448

### CODE REVIEW

**Comment by jreback** [comment]
@jorisvandenbossche this should take care of 1) and 3) in #12448 and doc 2)

**Comment by jreback** [comment]
@jorisvandenbossche ?

File: pandas/tseries/resample.py [hunk 1]
```diff
@@ -208,6 +208,12 @@ def _assure_grouper(self):
         """ make sure that we are creating our binner & grouper """
         self._set_binner()
 
+    def plot(self, *args, **kwargs):
+        # for compat with prior versions, we want to
+        # have the warnings shown here and just have this work
+        return _maybe_process_deprecations(self, how='mean').plot(*args,
```
Comment (jorisvandenbossche) [inline pandas/tseries/resample.py]: Small remark, by passing `how='mean'` here, the user gets the message "FutureWarning: how in .resample() is deprecated", while he did not use 'how' (in eg ``s.resample('15min').plot()``)

**Comment by jorisvandenbossche** [comment]
It is possible that the previous workaround with warning ("resample is now a deferred ..") for accessing elements does not work anymore?

```
In [13]: s.resample('15min').mean()[0]
Out[13]: -0.27773169409289944

In [14]: s.resample('15min')[0]
KeyError: 'Column not found: 0'
```

I thought this last one worked (gave the same result), but warned?

**Comment by jorisvandenbossche** [comment]
For case 3, I still not get the same as in 0.17.1. With this PR:

```
In [19]: df.groupby('group').resample('1D', fill_method='ffill')
C:\Anaconda\envs\devel\Scripts\ipython-script.py:1: FutureWarning: fill_method i
s deprecated to .resample()
the new syntax is .resample(...).ffill()
  if __name__ == '__main__':
Out[19]:
            val
date
2016-01-03    5
2016-01-10    6
2016-01-17    7
2016-01-24    8
```

So the actual grouper has disappeared

File: pandas/tseries/resample.py (hunk 1 above)
Comment (jreback) [inline pandas/tseries/resample.py]: no, he DID use ``mean`` implicity. Oh you want a more meaningful warning? ok sure

**Comment by jreback** [comment]
You get the same ``KeyError`` in 0.17.1 when you are using a ``DataFrame``

```
In [5]: s.to_frame('foo').resample('15T')[0]
```

I will fix the ``Series`` behavior

**Comment by jreback** [comment]
@jorisvandenbossche 

see #12486 

I can't fix this now. I *think* its possible, but will require some work. This is NOT the same as a multi-groupby of a column / freq, because resampling does some auto-filling. To be honest this is prob a bit to magical in the past.

**Comment by jreback** [comment]
@jorisvandenbossche any comments?

**Comment by jorisvandenbossche** [comment]
The error with the Series is still there (this is not necessarily related to the changes in this PR though):

```
In [21]: s.resample('2s').mean()[0]
Out[21]: 0.22864360896621477

In [22]: s.resample('2s')[0]
KeyError: 'Column not found: 0'
```

I think this should give something like iloc does?:

```
In [23]: s.resample('2s').iloc[0]
ValueError: .resample() is now a deferred operation
        use .resample(...).mean() instead of .resample(...)
        assignment will have no effect as you are working on a copy
```

**Comment by jorisvandenbossche** [comment]
BTW, the *"assignment will have no effect as you are working on a copy"* in that error message is not really clear to me. It's not that it has no effect, as it raises an error.

File: pandas/tseries/resample.py (hunk 1 above)
Comment (jorisvandenbossche) [inline pandas/tseries/resample.py]: Yep, that is what I mean (a more meaningful warning). As this is now rather confusing to the user IMO, as the warning says something about `how` while the user has not explicitely used it (before this PR, you got a "resample is now a deferred operation, please use `df.resample().mean() instead` message)

**Comment by jorisvandenbossche** [comment]
On the PR itself, plot change and whatsnew update looks good!

**Comment by jorisvandenbossche** [comment]
Something else with getitem:

```
In [8]: rs['2010-01-01 09:00:00']

AbstractMethodError: This method must be defined in the concrete class of Series
GroupBy
```

But these getitem things are not really related to this PR, so I can open another issue for that, and then this PR can be merged

**Comment by jreback** [comment]
Are you sure you did with this PR? (for the getitem issue). This is explicity fixed/mentioned.

```
In [5]: s.resample('2s').mean()[0]
Out[5]: -0.0026492975424703812

In [6]: s.resample('2s')[0]
Out[6]: -0.0026492975424703812
```

**Comment by jreback** [comment]
took out the assignment part of the message

**Comment by jreback** [comment]
ok, both those errors fixed (and now they show the deprecation warning as well).

ping if ok and i'll merge (I need to squash)

**Comment by jorisvandenbossche** [comment]
Looks good!

**Comment by jreback** [comment]
ok!

**Comment by benoit9126** [comment]
Hi,

I have a FutureWarning from pandas 0.18.0 related to a `resample` operation which seems inappropriate:
```python
import pandas as pd
import datetime as dt

df=pd.DataFrame(data=[1,3], index=[dt.timedelta(), dt.timedelta(minutes=3)])
df.resample('1T').interpolate(method='linear')
```
The result I get from the last line is correct but I get the following warning:
```
FutureWarning: .resample() is now a deferred operation use .resample(...).mean() instead of .resample(...)
```
What is wrong with my syntax?
Thanks

**Comment by jorisvandenbossche** [comment]
@benoit9126 No, I *think* this warning is correct, and I think what you want is: 

```
df.resample('1T').mean().interpolate(method='linear')
```

`interpolate` is not a method that is available on a Resample object. You first have to indicate how you want to resample the values (in this case using `mean` does work because you downsample).

But that aside, it does maybe make sense to have `interpolate` available, as this can work similarly to to `resample().fillna()`

**Comment by jreback** [comment]
@benoit9126 as @jorisvandenbossche points out, in < 0.18.0 there was an implicit ``how=mean`` being done when you said ``df.resample('1T)`` w/o any ``how=`` arg. 

@jorisvandenbossche yes, having a ``.interpolate`` method would be fine (and very straightforward to do actually). if you can open an issue with a nice example would be great.

**Comment by jorisvandenbossche** [comment]
@jreback Something else I just noticed, seems like a bug in ``asfreq``, but didn't yet test with master:

```
In [16]: df=pd.DataFrame(data=[1,3], index=[dt.timedelta(), dt.timedelta(minutes
=3)])
In [17]: df
Out[17]:
          0
00:00:00  1
00:03:00  3

In [18]: df.resample('1T').mean()
Out[18]:
            0
00:00:00  1.0
00:01:00  NaN
00:02:00  NaN
00:03:00  3.0

In [19]: df.resample('1T').asfreq()
Out[19]:
            0
00:00:00  1.0
00:01:00  NaN
00:02:00  NaN
```

Missing the the last value in asfreq?

**Comment by jreback** [comment]
@jorisvandenbossche hmm, that last does look like a bug, can you create an issue and i'll take a look. ty.

**Comment by benoit9126** [comment]
Thanks a lot for the explanation.
Nevertheless, I keep my code as it is hoping that it won't throw a warning in pandas 0.19 :wink:

**Comment by jorisvandenbossche** [comment]
@benoit9126 To be clear, there is no guarantee that #12925 will be implemented by 0.19, so it is possible that this code will start raising an error. Your responsibility :-) (or always welcome to put up a PR! I don't think it will be that hard to implement)

**Comment by benoit9126** [comment]
@jorisvandenbossche To be honest I do not have enough time to implement this feature in a close future (especially because I have never carefully inspected the pandas core). If it changes, it will be a pleasure.

**Pull request closed without merging**